compiler/sem/semfold.nim

```
@@ -774,4 +780,4 @@ proc getConstExpr(m: PSym, n: PNode; idgen: IdGenerator; g: ModuleGraph): PNode
     if i == n.len - 1:
       result = getConstExpr(m, n[i], idgen, g)
   else:
-    discard
+    discard # xxx: should we return nil for nkError?
```

`nil`, because it's only an expression or statement that can't be folded (which isn't an error in itself). E.g. an assignment.

---

Thinking about it a bit more, there are two situations involving `nkError`:
1. `nkError` is in the input, unless it's the root of the AST (base call of `getConstExpr`), this is probably an input violation because the caller failed to wrap the tree, but folding this shoud result in a `nil`
2. `nkError` is produced during a folding operation, these should be propagated and emitted

If a caller is speculatively folding via placeholder values and an error is generated, then the caller should discard these. Speaking of which, it's not entirely safe to do speculative folding over permuations of the same tree, because things like defines are effectful (practically it's rarely a problem, but hard to trust the code).

---

Given my above write-up, I think the current behaviour is correct, as all other occurances can be checked prior to dispatch to `getConstExpr` on parts of the sub-tree.